### PR TITLE
Added terminal ui library and demo

### PR DESCRIPTION
**Pull request by voldyman**
The terminal UI library lives in it's subdirectory and includes a demo for usage.

other reasons for change inline

### CODE REVIEW

File: src/commands/commands.cpp [hunk 1]
```diff
@@ -2,8 +2,7 @@
 #include "commands.h"
 #include "iostream"
 
-Commands::Commands() {
-    cmdList = {};
+Commands::Commands() : cmdList{} {
```
Comment (voldyman) [inline src/commands/commands.cpp:5]: made this change to fix builds with clang on a mac

File: src/term/terminal.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+#include <unistd.h>
+}
+
+const char* TERMINAL = "/dev/tty";
+
+struct TerminalImpl {
+    FILE *tty;
+    int fd;
+    struct termios oldTermios, newTermios;
+};
```
Comment (voldyman) [inline src/term/terminal.cpp:18]: Used the pimpl (private implementation pattern) to hide the C headers and unix specific includes from all the users of the library.

File: src/term/ui.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+int translateEscape(Terminal& term) {
+    char ch = term.getChar();
+    if (ch != '[') {
+        return ch;
+    }
+    ch = term.getChar();
+    if (ch == 'A' || ch == 'D') {
+        return Terminal::CTRL_P;
+    } else if (ch == 'B' || ch == 'C') {
+        return Terminal::CTRL_N;
```
Comment (voldyman) [inline src/term/ui.cpp:72]: emacs style movement keys, which are present by default in shells and libraries like readline.

File: src/uhh/CMakeLists.txt [hunk 4]
```diff
@@ -2,6 +2,8 @@ add_library(uhh-class uhh.cpp)
 
 list(APPEND EXTRA_LIBS stdc++fs)
 list(APPEND EXTRA_LIBS git2::git2)
+list(APPEND EXTRA_LIBS config-class)
+list(APPEND EXTRA_LIBS uhh-git)
```
Comment (voldyman) [inline src/uhh/CMakeLists.txt:6]: these were missing and were causing my builds on fail locally.

IMO we are going a little overboard with each class being it's own library, keeping uhh specific classes in one library might simplify the builds. Ninja takes care of incremental builds so we don't have to implement then ourselves.

**Comment by Toyz** [comment]
Looks good

File: src/uhh/CMakeLists.txt (hunk 4 above)
Comment (ThePrimeagen) [reply src/uhh/CMakeLists.txt:6]: Its simply the only way I know how to cmake :) 

Remember, i am a complete and utter rookie when it comes to cmake, programming, walking, and other life activities.

File: src/term/prompt.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        if (key == Terminal::DEL || key == Terminal::BACKSPACE) {
+            input.pop_back();
+        } else if (key == Terminal::CTRL_U) {
+            input.clear();
+        }
+        resetSelection = false;
+    } else {
+        resetSelection = false;
+    }
+    term.moveToColumn(1);
```
Comment (ThePrimeagen) [inline src/term/prompt.cpp:27]: ohh this is a nice interface.

**Review by ThePrimeagen — COMMENTED** [review]
I have never heard of the Pimpl, but apparently people disagree with how you defined your acronym.  https://www.geeksforgeeks.org/pimpl-idiom-in-c-with-examples/

**Comment by ThePrimeagen** [comment]
@voldyman I want to merge this but I would love a little walk through on this.

**Pull request merged**